hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/scheduler/ContainerScheduler.java

```diff
@@ -0,0 +1,393 @@
+              + 1);
+    }
+    container.sendLaunchEvent();
+  }
+
+  private List<Container> pickOpportunisticContainersToKill(
+      ContainerId containerToStartId) {
+    // The additional opportunistic containers that need to be killed for the
+    // given container to start.
+    List<Container> extraOpportContainersToKill = new ArrayList<>();
+    // Track resources that need to be freed.
+    ResourceUtilization resourcesToFreeUp = resourcesToFreeUp(
+        containerToStartId);
+
+    // Go over the running opportunistic containers.
+    // Use a descending iterator to kill more recently started containers.
+    Iterator<Container> reverseContainerIterator =
+        new LinkedList<>(scheduledToRunContainers.values()).descendingIterator();
+    while(reverseContainerIterator.hasNext() &&
+        !hasSufficientResources(resourcesToFreeUp)) {
+      Container runningCont = reverseContainerIterator.next();
+      if (runningCont.getContainerTokenIdentifier().getExecutionType() ==
+          ExecutionType.OPPORTUNISTIC) {
+
+        if (oppContainersMarkedForKill.containsKey(
+            runningCont.getContainerId())) {
+          // These containers have already been marked to be killed.
+          // So exclude them..
+          continue;
+        }
+        extraOpportContainersToKill.add(runningCont);
+        ResourceUtilizationManager.decreaseResourceUtilization(
+            getContainersMonitor(), resourcesToFreeUp,
+            runningCont.getResource());
+      }
+    }
+    if (!hasSufficientResources(resourcesToFreeUp)) {
+      LOG.warn("There are no sufficient resources to start guaranteed [{}]" +
+          "even after attempting to kill all running" +
+          "opportunistic containers.", containerToStartId);
+    }
+    return extraOpportContainersToKill;
+  }
+
+  private boolean hasSufficientResources(
+      ResourceUtilization resourcesToFreeUp) {
+    return resourcesToFreeUp.getPhysicalMemory() <= 0 &&
+        resourcesToFreeUp.getVirtualMemory() <= 0 &&
+        resourcesToFreeUp.getCPU() <= 0.0f;
+  }
+
+  private ResourceUtilization resourcesToFreeUp(
+      ContainerId containerToStartId) {
+    // Get allocation of currently allocated containers.
+    ResourceUtilization resourceAllocationToFreeUp = ResourceUtilization
+        .newInstance(this.utilizationManager.getCurrentUtilization());
+
+    // Add to the allocation the allocation of the pending guaranteed
+    // containers that will start before the current container will be started.
+    for (Container container : queuedGuaranteedContainers.values()) {
+      ResourceUtilizationManager.increaseResourceUtilization(
+          getContainersMonitor(), resourceAllocationToFreeUp,
+          container.getResource());
+      if (container.getContainerId().equals(containerToStartId)) {
+        break;
+      }
+    }
+
+    // These Resources have already been freed, due to demand from an
```

Comment:
Rephrase? "These resources are being freed, likely at the behest of another guaranteed container."?